README.rst
Carlo: describe a model and generate a list of events using this model
=====

    First of all he cut hairs on the log, then the forehead, the the eyes...

    Suddenly the eyes opened themselves and stared at him...

    Carlo gave no sign that he's frightened, just asked gently:

    – Wooden eyes, why do you look at me so strange?

    -- Aleksey Nikolaevich Tolstoy, "The Golden Key, or the Adentures of Buratino"

Although Papa Carlo was just a street organ player, his most famous achievement was creation of Buratino, speaking wooden doll. Using `carlo` you could also create (random) events following the described model.

Alpha version warning
---------------------

Current version is alpha only, all things may change in future: API, dependencies, and even the library's future is uncertain at this moment.

Example
-------

Image we want to create simple database containing information about photos of cats. Having two tables, ``cats`` and ``photos``, we want to generate **some amount of random data** for these tables. With ``carlo``, it could be done the following way::

    from carlo import *

    tables = Model(
        cats={
        'id': int_val(),
        'name': string_val(),
        'born': time_val(),
        },
        photos={
        'id': int_val(),
        'cat_id': int_val(),
        'taken': time_val(),
        }
    ).restricted_by(
        eq('cats.id', 'photos.cat_id'),
        ge('photos.taken', 'cats.born'),
        cardinal_many('cats', 'photos', 0, 10)
    ).build()

    generate(tables, 2)
    # which should produce:
    #
    # [('cats', {'id': 23424313434,
    #            'name': 'cat',
    #            'born': datetime(2001, 10, 13)}),
    #  ('photos', {'id': 49345,
    #              'cat_id': 23424313434,
    #              'taken': datetime(2004, 1, 12)}),
    #  ('photos', {'id': 3535209834399,
    #              'cat_id': 23424313434,
    #              'taken': datetime(2001, 12, 2)}),
    #  ('photos', {'id': 33444,
    #              'cat_id': 23424313434,
    #              'taken': datetime(2011, 6, 7)}),
    #  ('cats', {'id': 242431,
    #            'name': 'sdf sffd',
    #            'born': datetime(1912, 4, 12)}),
    # ]

Please note that **actual** state of the library is quite far from the declared in the example. Following things **do not work** at the current moment:

 * any restrictions except ``eq``
 * cardinality restrictions


NEWS.txt
News
====

0.0.7-alpha
-----

*Release date: UNRELEASED*

* The very first published version, not ready for use yet


test/test_event.py
#coding: utf-8
import os, sys
sys.path.append(os.path.join(sys.path[0], '..'))

from carlo import generate


class FakeModel(object):
    def __init__(self, *events):
        self.events = events
        self.i = -1

    def create(self):
        if self.i < len(self.events) - 1:
            self.i += 1
        else:
            self.i = 0
        return self.events[self.i]


def test_fake_model_returns_given_events_in_loop():
    m = FakeModel([3, 2, 1], [5, 4], [6])
    assert [3, 2, 1] == m.create()
    assert [5, 4] == m.create()
    assert [6] == m.create()
    assert [3, 2, 1] == m.create()
    assert [5, 4] == m.create()


def test_generate_one_shot():
    m = FakeModel([('a', {'x': 2}),
                   ('b', {'y': 3, 'z': 4})],
                  [('a', {'x': 5})],
                  [('a', {'x': 6}),
                   ('b', {'y': 7, 'z': 8}),
                   ('b', {'y': 9, 'z': 10})])
    assert [('a', {'x': 2}),
            ('b', {'y': 3, 'z': 4})] == \
           generate(m, 1)


def test_generate_multiple_shots():
    m = FakeModel([('a', {'x': 2}),
                   ('b', {'y': 3, 'z': 4})],
                  [('a', {'x': 5})],
                  [('a', {'x': 6}),
                   ('b', {'y': 7, 'z': 8}),
                   ('b', {'y': 9, 'z': 10})])
    assert [('a', {'x': 2}),
            ('b', {'y': 3, 'z': 4}),
            ('a', {'x': 5}),
            ('a', {'x': 6}),
            ('b', {'y': 7, 'z': 8}),
            ('b', {'y': 9, 'z': 10}),] == \
           generate(m, 3)


test/test_model.py
# coding: utf-8
import os, sys
sys.path.append(os.path.join(sys.path[0], '..'))

from carlo import Model, ModelException, eq, int_val, string_val
import pytest

def test_minimal_model():
    m = Model(const={'int': int_val(42)}).build()
    assert [('const', {'int': 42})] == m.create()
    m = Model(const2={'str': string_val('hello')}).build()
    assert [('const2', {'str': 'hello'})] == m.create()

def test_model_with_multiple_entities():
    m = Model(first={'name': string_val('elves')},
              second={'name': string_val('humans')}).build()
    assert sorted([('first', {'name': 'elves'}),
                   ('second', {'name': 'humans'})]) ==\
           sorted(m.create())

def test_model_with_multiple_params():
    m = Model(human={
        'head': int_val(1),
        'hands': int_val(2),
        'name': string_val('Hurin'),
        }).build()
    assert [('human', {'head': 1, 'hands': 2, 'name': 'Hurin'})] == m.create()

# restrictions

def test_restriction_must_override_parameter_definition():
    m = Model(leader={'direction': string_val('north')},
              follower={'direction': string_val()},
        ).restricted_by(eq('leader.direction', 'follower.direction')).build()
    assert sorted([('leader', {'direction': 'north'}),
                   ('follower', {'direction': 'north'})]) == \
           sorted(m.create())

def test_two_strings_of_same_length_could_be_eq():
    m = Model(x={'a': string_val(length=4),
                 'b': string_val(length=4)}
        ).restricted_by(eq('x.a', 'x.b')).build()
    row = m.create()[0][1]
    assert row['a'] == row['b']

# error handling
@pytest.mark.parametrize('model,restrictions', [
    ({'leader': {'direction': string_val('north')}, 'follower': {'direction': int_val(13)}},
     [eq('leader.direction', 'follower.direction')]),
    ({'x': {'a': string_val(length=5), 'b': string_val(length=4)}},
     [eq('x.a', 'x.b')]),
    ({'u': {'o': string_val('one'), 't': string_val('two')}},
     [eq('u.o', 'u.t')]),
    ({'fns': {'a': string_val(fn=lambda: 'hello'), 'b': string_val(fn=lambda: 'hello')}},
     [eq('fns.a', 'fns.b')]),
])
def test_failure_on_bad_combinations(model, restrictions):
    m = Model(**model).restricted_by(*restrictions)
    with pytest.raises(ModelException):
        m.build()


test/test_generators.py
# coding: utf-8
import os, sys
sys.path.append(os.path.join(sys.path[0], '..'))

from carlo import string_val
from carlo.generators import STR_TYPE
import pytest


def value_of(str_gen):
    return str_gen[0]()


class TestStringGenerator(object):
    def test_fixed_value(self):
        assert value_of(string_val('hello')) == 'hello'

    def test_list_of_fixed_values(self):
        chars = ['Harry', 'Ron', 'Hermiona']
        assert value_of(string_val(chars)) in chars

    def test_given_length(self):
        assert len(value_of(string_val(length=10))) == 10
        assert len(value_of(string_val(length=42))) == 42

    def test_given_length_and_prefix(self):
        val = value_of(string_val(length=15, prefix='so good '))
        assert len(val) == 15
        assert val.startswith('so good ')

        val = value_of(string_val(length=5, prefix='hello'))
        # len('hello') == 5
        assert val == 'hello'

    def test_incorrect_length_and_prefix(self):
        with pytest.raises(AssertionError):
            string_val(length=1, prefix='oh wait!')

    def test_single_prefix_without_length_is_forbidden(self):
        with pytest.raises(AssertionError):
            string_val(prefix='alone')

    def test_default_lengh_is_10(self):
        '''Because why not'''
        assert len(value_of(string_val())) == 10

    def test_implement_custom_generator_function(self):
        '''I do want easy Faker compatibility, but do NOT want have it in
        dependency'''
        val = value_of(string_val(fn=lambda: 'take this'))
        assert val == 'take this'


carlo/model.py
from .common import ModelException
import sympy


class Model(object):
    def __init__(self, **entities):
        self.entities = entities
        self.restrictions = list()

    def restricted_by(self, *args):
        self.restrictions.extend(args)
        return self

    def build(self):
        self._validate()
        return FrozenModel(self.entities, self.restrictions)

    def _validate(self):
        variables = dict()
        conditions = list()
        used_variabled = list()
        for name, params in self.entities.iteritems():
            for param_name, param in params.iteritems():
                full_name = '.'.join([name, param_name])
                variables[full_name] = dict()
                constraints = param[1]
                for k, v in constraints.iteritems():
                    var = sympy.symbols('.'.join([full_name, k]))
                    variables[full_name][k] = (var, v)
                    conditions.append(var - v)
                    used_variabled.append(var)
        for first, second in self.restrictions:
            for key in [k for k in variables[first] if k in variables[second]]:
                first_var = variables[first][key][0]
                second_var = variables[second][key][0]
                conditions.append(first_var - second_var)
        if conditions:
            ok = sympy.solve(conditions, used_variabled)
            if not ok:
                raise ModelException()


class FrozenModel(object):
    def __init__(self, entities, restrictions):
        self.entities = entities
        self.restrictions = {v:k for (k, v) in restrictions}

    def create(self):
        ready_values = dict()
        result = list()
        for name, params in self.entities.iteritems():
            for param_name, param in params.iteritems():
                full_name = '.'.join([name, param_name])
                ready_values[full_name] = param[0]()
        for name, params in self.entities.iteritems():
            resolved_params = dict()
            for param_name in params.keys():
                full_name = '.'.join([name, param_name])
                if full_name in self.restrictions:
                    full_name = self.restrictions[full_name]
                resolved_params[param_name] = ready_values[full_name]
            result.append((name, resolved_params))
        return result


carlo/generators.py
import random
import string


INT_TYPE = 1
STR_TYPE = 2


def int_val(fixed_value=None):
    return (lambda: fixed_value, {'type': INT_TYPE})


def string_val(fixed_value=None, length=None, prefix=None, fn=None):
    def rnd_str(str_len):
        return lambda: prefix + ''.join(random.choice(string.letters) for _ in range(str_len))

    constraints = {'type': STR_TYPE}
    if fixed_value:
        if isinstance(fixed_value, list):
            constraints['hash'] = hash(tuple(fixed_value))
            return (lambda: random.choice(fixed_value), constraints)
        constraints['hash'] = hash(fixed_value)
        return (lambda: fixed_value, constraints)
    if fn:
        constraints['unique'] = random.randint(0, 99999999999)
        return (fn, constraints)
    if prefix:
        assert length > 0
        length -= len(prefix)
        assert length >= 0
    else:
        prefix = ''
    if length is None:
        return (rnd_str(10), constraints)
    constraints['length'] = length
    return (rnd_str(length), constraints)


def time_val():
    pass


example.py
#!/usr/bin/env python2
# coding: utf-8

from carlo import *

tables = Model(
    cats={
     'id': int_val(),
     'name': string_val(),
     'born': time_val(),
    },
    photos={
     'id': int_val(),
     'cat_id': int_val(),
     'taken': time_val(),
    }
).restricted_by(
    eq('cats.id', 'photos.cat_id'),
    ge('photos.taken', 'cats.born'),
    cardinal_many('cats', 'photos', 0, 10)
).build()

generate(tables, 2)
# [('cats', {'id': 23424313434,
#            'name': 'cat',
#            'born': datetime(2001, 10, 13)}),
#  ('photos', {'id': 49345,
#              'cat_id': 23424313434,
#              'taken': datetime(2004, 1, 12)}),
#  ('photos', {'id': 3535209834399,
#              'cat_id': 23424313434,
#              'taken': datetime(2001, 12, 2)}),
#  ('photos', {'id': 33444,
#              'cat_id': 23424313434,
#              'taken': datetime(2011, 6, 7)}),
#  ('cats', {'id': 242431,
#            'name': 'sdf sffd',
#            'born': datetime(1912, 4, 12)}),
# ]
